proposals/0000-overloaded-bracket.rst

```diff
@@ -0,0 +1,394 @@
+ambiguity in the composition.
+
+It doesn't seem to me that there will be any problems with ambiguity here as
+the types of splices is not overloaded in the same manner.
+
+
+Interaction with Lift
+.....................
+
+The main breakage from this patch comes from modifying the type signature for
+``lift``.
+
+Instances defined using ``DeriveLift`` will continue to work because they are
+defined in terms of quotation brackets.
+
+Instances written in terms of the combinators from ``Language.Haskell.TH.Lib``
+will continue to work because these combinators will be generalised.
+
+Instances written in terms of ``Q`` will no longer work. For users to migrate
+an additional class ``LiftQ`` could be defined which has the old interface.
+This would mean users need to explicitly lift but there are likely only a few
+instances which fall into this category if any at all.
+
+Definition of Quote
+...................
+
+Richard observes that ``Language.Haskell.TH.Lib.Internal.numTyLit`` calls
+``fail`` from the ``Q`` monad. This call to ``fail`` can be replaced with
+a call to ``error``. It will still be executed at compile-time but with a
+potentially slightly worse error message. The alternative is to
+also add this effect to the ``Quote`` type class.
+
+Other Effects
+.............
+
+Vlad points out that you don't need to very strict about the types of
+expressions in splices. Each nested splice could have different constraints::
+
+      f :: Quasi m => m Exp
+      g :: MonadIO m => m Exp
+      [| putStrLn $(f) >> putStrLn $(g) |] :: (Quote m, Quasi m, MonadIO m) => m Exp
+
+If one of the nested splices has a specific type, for instance ``Q Exp``, then
+the type of the whole expression is fixed to be ``Q Exp``.
+
+
+Costs and Drawbacks
+-------------------
+
+* The generalisation of untyped brackets does not seem like it will cause
+  any significant breakage but it's hard to predict.
+* The modification to the ``Lift`` interface could cause user-written instances
+  to break but users should not define their own instances anyway.
+  ``DeriveLift`` is the blessed manner in which to define a ``Lift`` instance.
+
+Alternatives
+------------
+
+* The main alternative to the design would be to only require a ``Quote``
+  constraint when the quotation requires the ``newName`` effect. For example,
+  ``[| 5 |] :: Applicative m => m Exp``. I am opposed to this direction as it
+  breaks abstraction. The implementation detail of how ``[| 5 |]`` is desugared
+  leaks to the user.
+
+Unresolved Questions
+--------------------
+
+* Carter points out that if you want to achieve "pure" template haskell then
+  you still need to deal with the fact that different platforms have different
+  representations of primitive data types. This is out of scope of this
+  proposal.
+
+* It would also be possible to make ``Quote`` a superclass of ``Quasi`` but
+  this hierarchy refactoring seems unecessary.
```

Comment:
On the other hand, is there any situation where we may want ``Quasi`` without ``Quote``?

Reply:
I'm not aware of anyone definition a `Quasi` instance for themselves so it seems difficult to answer.

Reply:
I've hand-written `Quasi` instances profitably. See https://github.com/goldfirere/singletons/blob/master/src/Data/Singletons/Single/Monad.hs#L55. And I've used GeneralizedNewtypeDeriving twice: https://github.com/goldfirere/singletons/blob/master/src/Data/Singletons/Promote/Monad.hs#L49 and https://github.com/goldfirere/th-desugar/blob/master/Language/Haskell/TH/Desugar/Reify.hs#L206. So it's possible.